File: scripts/bench_inverse.py
"""Time the seven inverse expansions against Sage: the four Macdonald ones
`s -> Htilde`, `s -> J`, `m -> P`, `m -> Q`, and the three Jack ones
`m -> P`, `m -> Q`, `m -> J`.

    python scripts/bench_inverse.py [top_degree]

The unit of work is the **degree**: every partition of `n`, expanded out of the
Schur basis into the parametric one. That is what a caller asking whether some
family of functions is `Htilde`- or `J`-positive does, and it is the unit both
sides amortize their transition matrix over.

Each degree and each arm runs in its **own process**, and the arm part of that
is not optional: Sage caches the transition matrices behind these conversions,
the cache is not confined to one degree, and it is *shared between the
normalizations of one family*. Two Jack arms in one process read 20x faster
than they do alone, because the first one built the matrix the other two then
reused (`docs/record/jack.md`). symfn memoizes too; `clear_caches` in the
example drops it between workloads.

The Sage arm needs `SAGE_DISABLE_SYMFN` in its *environment*, and this script
refuses to run without it taking effect: Sage's Macdonald bases reach this
library through the optional backend, so an arm with it enabled compares symfn
to symfn and reports about 1.0x.

The two families print as two tables rather than one wide row, and they run
over different base rings: the Macdonald arms over the fraction field of
`QQ[q,t]`, the Jack ones over `QQ[alpha]`'s. Building the ring is outside the
timed region on both sides.

Sage is a black box here: run and timed, never read (see NOTICE.md).

WARNING: Record the power state with the numbers. This machine drifts about
1.8x on battery, and it changes ratios and not only times.
"""

import os
import subprocess
import sys
import time


def one(n, arm):
    from sage.all import Partitions, PolynomialRing, QQ, SymmetricFunctions

    from sage.libs.symfn import is_available

    if is_available():
        print(
            "the symfn backend is enabled; run with SAGE_DISABLE_SYMFN=1 in the "
            "environment or this compares symfn to symfn",
            file=sys.stderr,
        )
        sys.exit(1)

    if arm < 4:
        R = PolynomialRing(QQ, "q,t").fraction_field()
        Sym = SymmetricFunctions(R)
        mac = Sym.macdonald()
        basis = (mac.Ht(), mac.J(), mac.P(), mac.Q())[arm]
        source = Sym.schur() if arm < 2 else Sym.monomial()
    else:
        A = PolynomialRing(QQ, "a").fraction_field()
        Sym = SymmetricFunctions(A)
        jack = Sym.jack(t=A.gen())
        basis = (jack.P(), jack.Q(), jack.J())[arm - 4]
        source = Sym.monomial()

    # One degree-1 conversion first, untimed: Sage builds a family's coercion
    # machinery on its first use, and without this the arm that ran first paid
    # for the whole family. It touches no transition matrix of the degree
    # being measured.
    basis(source(Partitions(1)[0]))

    shapes = list(Partitions(n))
    start = time.perf_counter()
    for la in shapes:
        basis(source(la))
    print(len(shapes), f"{time.perf_counter() - start:.4f}")


if len(sys.argv) > 3 and sys.argv[1] == "--one":
    one(int(sys.argv[2]), int(sys.argv[3]))
    sys.exit(0)

TOP = int(sys.argv[1]) if len(sys.argv) > 1 else 6
ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

env = dict(os.environ)
env["SAGE_DISABLE_SYMFN"] = "1"

#: The Sage arm's seven timings, in order, against the example's workload
#: names, split into the two tables they print as.
MACDONALD = (("s->Ht", "ht"), ("s->J", "j"), ("m->P", "p"), ("m->Q", "q"))
JACK = (("m->P", "jack_p"), ("m->Q", "jack_q"), ("m->J", "jack_j"))
ARMS = MACDONALD + JACK


def ratio(theirs, ours):
    return f"{theirs / ours:>6.1f}x" if ours > 0 else "      --"


def header(arms):
    out = f"{'n':>3} {'p(n)':>5}"
    for label, _ in arms:
        out += f" {label + ' sage':>12} {'symfn':>9} {'ratio':>7}  "
    return out


rows = []
for n in range(1, TOP + 1):
    sage = []
    for arm in range(len(ARMS)):
        theirs = subprocess.run(
            [sys.executable, os.path.abspath(__file__), "--one", str(n), str(arm)],
            capture_output=True,
            text=True,
            env=env,
            cwd=ROOT,
        )
        if theirs.returncode:
            sys.stderr.write(theirs.stderr)
            sys.exit(1)
        count, secs = theirs.stdout.split()
        sage.append(float(secs))

    cargo = ["cargo", "run", "--release", "--quiet", "--example", "bench_inverse"]
    mine = subprocess.run(
        [*cargo, "--", str(n)], capture_output=True, text=True, cwd=ROOT
    ).stdout.split()
    ours = {mine[i]: float(mine[i + 1]) for i in range(0, len(mine), 3)}

    rows.append((n, count, sage, ours))

for title, arms in (("Macdonald", MACDONALD), ("Jack", JACK)):
    print(f"\n{title}")
    print(header(arms))
    for n, count, sage, ours in rows:
        row = f"{n:>3} {count:>5}"
        for label, name in arms:
            theirs_secs = sage[[a[1] for a in ARMS].index(name)]
            row += (
                f" {theirs_secs:>11.4f}s {ours[name]:>8.4f}s "
                f"{ratio(theirs_secs, ours[name]):>7}  "
            )
        print(row)

File: examples/bench_inverse.rs
//! Time the inverse expansions `s → H̃`, `s → J`, `m → P`, `m → Q` and the
//! three Jack ones, one degree per process.
//!
//! ```text
//!   cargo run --release --example bench_inverse -- 6
//! ```
//!
//! Prints `<workload> <seconds>` for one degree and exits. The degree is the
//! unit and the process is the boundary: both this crate and Sage memoize
//! across degrees, and `docs/record/qt-kostka.md` measures what a sequential
//! run does to the numbers. `scripts/bench_inverse.py` drives one process per
//! degree and per arm.
//!
//! The workload is **every** λ of the degree, which is what a caller asking
//! whether some family of functions is `H̃`- or `J`-positive does.
//! `j_table` is the same answer from the whole-degree entry point, which is
//! where a caller with every shape of one degree should go. `p`, `q` and the
//! three `jack_*` workloads take the monomial basis rather than the Schur
//! one, because that is what those expansions are written in.

use std::time::Instant;

use symfn::{
    clear_caches, monomial_to_jack_j, monomial_to_jack_p, monomial_to_jack_q,
    monomial_to_macdonald_p, monomial_to_macdonald_q, partitions_of, schur_in_j_table,
    schur_to_macdonald_ht, schur_to_macdonald_j, AFrac, Frac, Monomial, QtPoly, Rational, Ring,
    Schur, SymFn,
};

fn s_lambda(lambda: &symfn::Partition) -> Schur<QtPoly<Rational>> {
    Schur::monomial(lambda.clone(), QtPoly::term(0, 0, Rational::from_int(1)))
}

fn m_lambda(lambda: &symfn::Partition) -> Monomial<Frac<Rational>> {
    Monomial::monomial(lambda.clone(), <Frac<Rational> as Ring>::one())
}

fn m_alpha(lambda: &symfn::Partition) -> Monomial<AFrac<Rational>> {
    Monomial::monomial(lambda.clone(), <AFrac<Rational> as Ring>::one())
}

fn main() {
    let n: u32 = std::env::args()
        .nth(1)
        .and_then(|s| s.parse().ok())
        .unwrap_or(5);

    let jobs: Vec<(&str, fn(u32) -> usize)> = vec![
        ("ht", |n| {
            partitions_of(n)
                .iter()
                .map(|l| schur_to_macdonald_ht(&s_lambda(l)).len())
                .sum()
        }),
        ("j", |n| {
            partitions_of(n)
                .iter()
                .map(|l| schur_to_macdonald_j(&s_lambda(l)).len())
                .sum()
        }),
        ("j_table", |n| {
            schur_in_j_table::<Rational>(n)
                .iter()
                .flatten()
                .filter(|c| !Ring::is_zero(*c))
                .count()
        }),
        ("p", |n| {
            partitions_of(n)
                .iter()
                .map(|l| monomial_to_macdonald_p(&m_lambda(l)).len())
                .sum()
        }),
        ("q", |n| {
            partitions_of(n)
                .iter()
                .map(|l| monomial_to_macdonald_q(&m_lambda(l)).len())
                .sum()
        }),
        ("jack_p", |n| {
            partitions_of(n)
                .iter()
                .map(|l| monomial_to_jack_p(&m_alpha(l)).len())
                .sum()
        }),
        ("jack_q", |n| {
            partitions_of(n)
                .iter()
                .map(|l| monomial_to_jack_q(&m_alpha(l)).len())
                .sum()
        }),
        ("jack_j", |n| {
            partitions_of(n)
                .iter()
                .map(|l| monomial_to_jack_j(&m_alpha(l)).len())
                .sum()
        }),
    ];

    for (name, f) in jobs {
        // Cold, so no workload is charged for another's tables.
        clear_caches();
        let t0 = Instant::now();
        let terms = f(n);
        println!("{name} {:.4} {terms}", t0.elapsed().as_secs_f64());
    }
}
